libfwupd/fwupd-client.c

```diff
@@ -428,7 +366,90 @@ fwupd_client_connect (FwupdClient *self, GCancellable *cancellable, GError **err
 	if (val != NULL)
 		fwupd_client_set_host_security_id (self, g_variant_get_string (val, NULL));
 
-	return TRUE;
+	/* success */
+	g_task_return_boolean (task, TRUE);
+}
+
+static void
+fwupd_client_connect_get_bus_cb (GObject *source,
+				 GAsyncResult *res,
+				 gpointer user_data)
+{
+	g_autoptr(GTask) task = G_TASK (user_data);
+	FwupdClient *self = g_task_get_source_object (task);
+	FwupdClientPrivate *priv = GET_PRIVATE (self);
+	g_autoptr(GError) error = NULL;
+
+	priv->conn = g_bus_get_finish (res, &error);
+	if (priv->conn == NULL) {
+		g_prefix_error (&error, "Failed to connect to system D-Bus: ");
+		g_task_return_error (task, g_steal_pointer (&error));
+		return;
+	}
+	g_dbus_proxy_new (priv->conn,
+			  G_DBUS_PROXY_FLAGS_NONE,
+			  NULL,
+			  FWUPD_DBUS_SERVICE,
+			  FWUPD_DBUS_PATH,
+			  FWUPD_DBUS_INTERFACE,
+			  g_task_get_cancellable (task),
+			  fwupd_client_connect_get_proxy_cb,
+			  g_object_ref (task));
+}
+
+/**
+ * fwupd_client_connect_async:
+ * @self: A #FwupdClient
+ * @cancellable: the #GCancellable, or %NULL
+ * @callback: the function to run on completion
+ * @callback_data: the data to pass to @callback
+ *
+ * Sets up the client ready for use. This is probably the first method you call
+ * when wanting to use libfwupd.
+ *
+ * Since: 1.5.0
+ **/
+void
+fwupd_client_connect_async (FwupdClient *self, GCancellable *cancellable,
```

Comment:
I'm not sure an async design you want a separate connect method.  By definition connecting is supposed to be synchronous.  If this doesn't finish but you've already called something that expectes you to have connected isn't that a problem?

Reply:
It forces initialization into a three step process (which is actually a good thing). Let me explain a bit:

* phase 1 is app startup, loading GUIs etc
* phase 2 is contacting fwupd and connecting the proxy -- showing an error dialog if not available
* phase 3 is using the proxy and firing off all the requests async to be completed whenever

You can see all this in action in https://gitlab.gnome.org/hughsie/gnome-firmware-updater/-/blob/wip/hughsie/async/src/gfu-main.c#L1455

Reply:
Ahhh.. seeing `fwupd_client_connect_finish` in action that flow makes a lot more sense.

Why not also bring fu-util into the same approach?   I guess maybe not as valuable since there isn't a GUI to click and interact that gets blocked while waiting for stuff to get done.

Reply:
> that flow makes a lot more sense

I've added some more doctext to the fwupd_client_connect_async() method making this a bit more clear.

> Why not also bring fu-util into the same approach